Add STM32 cordic support

qosch opened the pull request:
This implements functions to use the CORDIC peripheral present in newer STM32 devices (STM32G4 and STM32U5). It is compiling and I have tested the 16 and 32 cosine functions on STM32g4 hardware. DMA and interrupts as well as the convenience functions for sine still needs testing, but I expect no major changes (since sine is no different than cosine and DMA and interrupt enable functions just set one bit). So for me this is ready for review.

karlp commented on include/libopencm3/stm32/common/cordic_common_v1.h:
```diff
@@ -0,0 +1,159 @@
+
+/**@{*/
+
+/* THIS FILE SHOULD NOT BE INCLUDED DIRECTLY, BUT ONLY VIA ADC.H
```
typo

karlp commented on lib/stm32/common/cordic_common_v1.c:
```diff
@@ -0,0 +1,427 @@
+ *
+ * Use this function to set write 16 bit arguments to the 32 bit CORDIC_WDATA register.
+ * @param[in] argument1 Unsigned int16 argument1
+ * @param[in] argument2 Unsigned int16 argument2
```
you don't need the "Unsigned int16" stuff, that's obvious from the signatures...

qosch replied:
I removed them from all @param but the parameter descriptions became kind of useless. Shall I drop them completely?

karlp commented on lib/stm32/common/cordic_common_v1.c:
```diff
@@ -0,0 +1,427 @@
+        cordic_set_16bit_result_data_width();
+        cordic_set_number_of_arguments_1();
+        cordic_set_number_of_results_1();
+        // scale is not applicable for cos
```
only us C style comments throughout please

qosch commented on lib/stm32/common/cordic_common_v1.c:
```diff
@@ -0,0 +1,427 @@
+
+        // this while loop can be omitted but that will stall the 
+        // processor while it waits for the CORDIC_RDATA register
+        while(!cordic_get_result_ready_flag());
```
These functions are not really suitable for time-critical code anyway. The reference manual describes how to use the peripheral instead: configure it once during setup and only write the argument(s) on every cycle of e.g. control loops. Then perform something else while cordic does its thing and read the result a few clock cycles later.
I put "blocking" in the brief description but would not provide a non blocking version. I hope the hint in the description is clear enough.

karlp replied:
You don't seem to have any way of using the apis you provided to do that though?  The only methods you've provided for doing cos/sin() is via blocking calls?  What am I missing?

qosch replied:
I don't think there is much to provide. This is what an application with a setup phase and a fast real time control loop (the `while(true)` loop) that needs the cos of an `int16_t x;` would look like:

```
void cordic_setup(void) {
	cordic_set_function(CORDIC_CSR_FUNC_COS);
	cordic_set_precision(CORDIC_CSR_PRECISION_ITER_20);
	cordic_set_argument_width_16bit();
	cordic_set_result_width_16bit();
	cordic_set_number_of_arguments_1();
	cordic_set_number_of_results_1();
	/* scale is not applicable for cos */
}

void main(void) {
	/* setup everything else */
	cordic_setup();

	while(true) {
		/* real time control loop */
		/* do something before */
		cordic_write_16bit_arguments((uint16_t) x, 0x7FFF);
		/* use the free processor time while cordic peripheral does its thing */
		/* if the result isn't ready by the time we reach the cordic_read function, 
		the read will take as long as the peripheral needs to finish. */
		y = cordic_read_16bit_result();
		/* do something after */
	}
} 
```
So once the peripheral is configured, only the `cordic_write()` and `cordic_read()` functions (in its different 16 and 32bit flavors) are needed. I could of course split up the setup part of the functions provided so far (the first 7 lines of `cordic_cos_16bit`, `cordic_cos_32bit`, `cordic_sin_16bit` and `cordic_sin_32bit`) into four functions called `cordic_configure_for_cos_16bit` and the like.

qosch commented:
Concerning G0: I initially thought G0 had a cordic peripheral but it doesn't seem to (cant find it in the reference manual). I hope I removed it everywhere, now. I added a device header for U5 already, though.

qosch commented on lib/stm32/common/cordic_common_v1.c:
```diff
@@ -0,0 +1,427 @@
+ *
+ * Use this function to write a 32 bit argument to the CORDIC_WDATA register. 
+ * If the operation needs two arguments call cordic_set_number_of_arguments_2() 
+ * before and then use this function twice to write both arguments.
```
Yes that's right. I am not really familiar with the term "word", is it the same width independent of the exact platform? I changed the function name a few times until I decided I cant find a name that explains the behavior. If your proposal is self-explanatory to you I'll take it.

karlp commented:
So, just so I understand, there's still no methods in your API for starting a cos/sin function and checking later?  you would have to manually build that up yourself?  You've added the function for checking for the result being ready, but still only have blocking versions of the trig functions?

qosch commented:
after initialization of the peripheral, the only functions needed for that are
`cordic_write_16bit_arguments();`
and 
`cordic_read_16bit_result();` (or the corresponding 32bit version).
These functions will perform whatever operation the peripheral was configured for. Writing to the argument register will trigger the operation without any "start" or "enable" bit. Reading the result register is blocking, but there is no way around that.
I'll provide `cordic_configure_for_cos_16bit`, `cordic_configure_for_cos_32bit`, `cordic_configure_for_sin_16bit` and `cordic_configure_for_sin_32bit` so that using the peripheral in an efficient manner will only take three function calls: one of the config functions during initialization and the write and read functions every time an operation is to be performed.
The relevant part is on p.478 "Zero-overhead mode" of the reference manual (https://www.st.com/resource/en/reference_manual/rm0440-stm32g4-series-advanced-armbased-32bit-mcus-stmicroelectronics.pdf).
Maybe I misunderstood but I don't think a non blocking functions that reconfigures the peripheral on every call is very helpful because it is unnecessary in most applications.

qosch commented:
I've merged latest changes from master, added functions to configure the peripheral for a few modes as well as non blocking versions of the functions that were present before.

karlp commented:
I've merged this via rebase in 5db7e87e (I dropped the non-existant u5, it's easy to put back when we have u5 code)  thanks!
